src/fuse/file_system.cr

```diff
@@ -100,11 +105,22 @@ module Fuse
 
       @operations.read = ->(path : LibC::Char*, buf : LibC::Char*, size : LibC::SizeT, offset : LibC::OffT, fi : Binding::FileInfo*) do
         r = invoke_file read, path, fi, buf.as(UInt8*).to_slice(size), offset
-
         return r if r.is_a?(Int32)
         result r
       end
 
+      @operations.readlink = ->(path : LibC::Char*, buf : LibC::Char*, size : LibC::SizeT) do
+        r = invoke readlink, String.new(path)
+        if r.nil?
+          -LibC::ENOSYS
+        else
+          buffer = buf.as(UInt8*).to_slice(size)
+          target = r + "\0"
```

Comment:
I'm not sure, but weren't Crystal Strings guaranteed to be NUL-terminated anyway?

Reply:
oh? Interesting; i didn't know that but I'm going to look it up and chance the code accordingly.

Thanks for the thorough review!

Reply:
I couldn't find any documentation on this but I tried it out and it appears not the be the case that we get NUL-terminated strings. So in the end I shuffled some lines around but the code functionally stayed the same. (And we learned something! :smile: )

You can easily try this for yourself if you like of course:

```crystal
"foo".to_slice.inspect
"foo\0".to_slice.inspect # to verify that inspect is not hiding the 0-byte; one cannot be to careful
```

Reply:
`String#to_slice` doesn't give you the NUL as that's more convenient.  It's there however:

```cr
p "foo".to_unsafe.to_slice(4).inspect #=> Bytes[102, 111, 111, 0]
p "#{123}".to_unsafe.to_slice(4).inspect #=> Bytes[49, 50, 51, 0]
```

It's been a long time since I've dug into Crystal, so I don't remember anymore where it's done.  You should be able to find the latter one in the `String::Builder` class, as that is instantiated to process the `#{}` interpolation.  iirc it's done for C API compatibility - Which is our use-case.

Reply:
Interesting! I updated the code again since you are obviously (observably) correct. Using bytesize + 1 when I cannot find any mention that Crystal guarantees that the \0 will always be there makes me feel quite uneasy though even though it's very likely that it will. Ah well; it's probably okay.

Reply:
Story of my life :shrug: